Derive path-based package names in a file system agnostic way

jonas opened the pull request:
@tindzk, @muxanick Based on your discussion [here](https://github.com/scala-native/scala-native/pull/691#discussion_r115148711), I'm proposing this change as a way to refactor the three versions that convert relative paths to package names into a utility method which is file system (and file separator) agnostic. This avoids using regex by leveraging [`java.nio.file.Path#iterator`](https://docs.oracle.com/javase/7/docs/api/java/nio/file/Path.html#iterator()) to iterator over each path segment.

ekrich commented:
Since we are working inside sbt there are utilties that help deal with paths written in Scala. http://www.scala-sbt.org/0.13/docs/Paths.html The `/` is a methods that puts the correct separator for the platform. This is part of the sbt IO library that is used by sbt itself and people making build files. https://github.com/sbt/io This way you don't have to reinvent stuff. Main developers, Eugene and Dale are on this channel and do help. https://gitter.im/sbt/sbt

muxanick commented:
@jonas good stuff, there is also Nir codegen which uses both slashes, I'm not sure what to do with it.
@ekrich does that mean if we are going to use Paths then we will be bound to sbt? I have a plan to create scala-nativec.exe which I will invoke from IDE without any sbt.

ekrich commented:
@muxanick It is a separate dependency - assuming has no external dependencies and it supports 2.10, 2.11, and 2.12. It does seem to use some Java things we don't support at the minute but not sure if that would prevent it from working or not. https://mvnrepository.com/artifact/org.scala-sbt/io_2.11/1.0.0-M1

jonas commented:
I'm open to rework the PR to use sbt-io, but since Scala Native requires Java 1.7 I don't see much advantage since equivalents to`sbt.IO.{relativize,pathSplit}` are provided by the NIO API. The tools project and therefore also the util project are pulled into the compiler plugin, so less dependencies seems better.

ekrich commented:
Since Scala Native is a relatively new project, it would make sense to have a good strategy forward. Java 1.7 was left for one person I think - it was tried to force 1.8 but backed off. Some decisions were made based on sbt and Scala version when the project was started and I don't think anyone wants to disrupt everything at this point. 

Less dependencies are always better unless you end up writing a lot of code and the dependency is good and maintained. The IO stuff is included in SBT 0.13 but in 1.0 it is an external dependency. I'm not sure what is the best path, was just making suggestions.

muxanick commented:
I would prefer less code and less dependencies and I don't mind sbt-io or NIO, latter looks more general though, I personally prefer some abstract front-end which allows easily to replace back-end.

ekrich commented:
A couple of points to consider. Not trying to be pushy but have been following sbt for quite a while.

- If you are using sbt for the plugin or the build itself then using sbt.io makes sense as that is what most people use to make things easier in a build or plugin.
- sbt 1.0 which is coming out soon will also use sbt.io as a dependency and is also published for others to use. sbt 1.0 is Scala 2.12 and Java 8.
- If we have other needs or improvements we could submit PRs to sbt.io
- sbt team is supported by Lightbend and maintains sbt.io

@muxanick I can't exactly know what you have in mind but if this initiative was just for sbt and/or the native plugin then sbt.io would be a safe bet.

muxanick commented:
@ekrich if I understand correctly both use java.io.File underhood and both are independent, so no big difference for me. I would love to have abstraction from path separator.
Does sbt.io require some additional work?

jonas commented:
@ekrich It would be helpful if you could be more specific regarding which part of sbt-io you think are better suited than the Java 7 APIs used by this PR. I've converted places in `build.sbt` to use `sbt.IO.relativizeFile`. To my knowledge sbt-io doesn't provide a way to convert a file path to a package name, and using `Path` seems cleaner than `sbt.IO.pathSplit` which relies on a regex under the hood.

@muxanick I think d5ade13 addresses the path issue in NirCodeGen (and NirFiles) you mentioned.

tindzk commented on tools/src/main/scala/scala/scalanative/linker/Classpath.scala:
```diff
@@ -19,22 +19,20 @@ sealed trait Path {
 }
 
-object Path {
+object Classpath {
```
How about writing `ClassPath` instead? Then it would conform to the spelling of `scala.tools.nsc.util.ClassPath`.

ekrich replied:
@jonas Using Java 7 APIs is fine or sbt io utilities as convenient. I was just trying to point out the state of affairs with the sbt versions etc. The sbt folks are part of the same community so reach out if needed. You guys are doing great stuff so carry on.

jonas replied:
Done in ce52ad0

tindzk commented on util/src/main/scala/scala/scalanative/io/package.scala:
```diff
@@ -14,4 +16,13 @@ package object io {
     finally pool.reclaim(buffer)
   }
+
+  def packageNameFromPath(path: Path, extension: String): String = {
```
It seems to me that the `extension` parameter is redundant. What do you think about this?

```scala
  def packageNameFromPath(path: Path): String = {
    val fileName = path.getFileName.toString
    val base     = fileName.split('.').init.mkString(".")

    Option(path.getParent) match {
      case Some(parent) => parent.resolve(base).asScala.mkString(".")
      case None         => base
    }
  }
```

jonas replied:
Yes, that looks better. Fixed in 69b4328

tindzk commented:
Great work! This is a nice solution.

As for the discussion about including the sbt-io dependency, I am rather sceptical. As the name suggests, its functionality satisfies the needs of sbt. Given that there is no documentation, I would think of it as a helper library for sbt, not meant for public use.

If we wanted to rely on external functionality, it must be a more more generic library. Generally, we should strive to reduce dependencies and only introduce them if they really give us a benefit. Each additional dependency may hold us back in the future for various reasons (regressions, performance issues, etc.).

ekrich commented:
@jonas You guys are doing great work. I wanted to just let you know the state of affairs with sbt. The sbt team is part of the community and willing to help and make sbt better. Here is a long thread.
https://contributors.scala-lang.org/t/asking-for-your-feedback-on-sbt-scala-center-announcement/738/1

@tindzk The version of sbt we are using, IO is an internal helper library but it has been made standalone for sbt 1.0 precisely so it can change independent of sbt.

muxanick commented:
@jonas Hey Jonas, any news about this pull? I desperately need these changes to finish my work on windows port.

jonas commented:
@muxanick We need to wait for @densh to have some free time to review the PR. I've rebased it to trigger a build so at least that won't block the review.

Depending on how comfortable you are with git you can
```
git pull git@github.com:jonas/scala-native.git path-refactoring
```
to merge the changes into your windows port.

muxanick commented:
@jonas Thanks! I will try

muxanick commented:
I've tried your changes and it works well for me, but I still have same issue with 'tools/tests':
[info] scala.scalanative.FrameworkTest *** ABORTED ***
[info]   java.lang.NoClassDefFoundError: scala/collection/Seq
[info]   at java.lang.Class.getDeclaredConstructors0(Native Method)
[info]   at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
[info]   at java.lang.Class.getConstructor0(Class.java:3075)
[info]   at java.lang.Class.getConstructor(Class.java:1825)
[info]   at scala.scalanative.NIRCompiler$.getCompiler(NIRCompiler.scala:56)
[info]   at scala.scalanative.LinkerSpec$$anonfun$link$1.apply(LinkerSpec.scala:35)
[info]   at scala.scalanative.LinkerSpec$$anonfun$link$1.apply(LinkerSpec.scala:33)
[info]   at scala.scalanative.util.Scope$.apply(Scope.scala:30)
[info]   at scala.scalanative.LinkerSpec.link(LinkerSpec.scala:33)
[info]   at scala.scalanative.FrameworkTest$$anonfun$1.apply(FrameworkTest.scala:14)
[info]   ...
[info]   Cause: java.lang.ClassNotFoundException: scala.collection.Seq
[info]   at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
[info]   at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
[info]   at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
[info]   at java.lang.Class.getDeclaredConstructors0(Native Method)
[info]   at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
[info]   at java.lang.Class.getConstructor0(Class.java:3075)
[info]   at java.lang.Class.getConstructor(Class.java:1825)
[info]   at scala.scalanative.NIRCompiler$.getCompiler(NIRCompiler.scala:56)
[info]   at scala.scalanative.LinkerSpec$$anonfun$link$1.apply(LinkerSpec.scala:35)
[info]   at scala.scalanative.LinkerSpec$$anonfun$link$1.apply(LinkerSpec.scala:33)

any ideas?

jonas commented:
@muxanick Since this looks like a classpath problem, try this path. In `build.sbt` those classpaths are created using `File.pathSeparator`

```diff
diff --git i/testing-compiler/src/main/scala/scalanative/compiler/NIRCompiler.scala w/testing-compiler/src/main/scala/scalanative/compiler/NIRCompiler.scala
index 5c674d9..85f454d 100644
--- i/testing-compiler/src/main/scala/scalanative/compiler/NIRCompiler.scala
+++ w/testing-compiler/src/main/scala/scalanative/compiler/NIRCompiler.scala
@@ -83,7 +83,7 @@ class NIRCompiler(outputDir: File) extends api.NIRCompiler {
                                val classpath: List[String])
       extends CompilerOption(
         s"-Xplugin:$jarPath" + (if (classpath.nonEmpty)
-                                  classpath.mkString(" -cp ", ":", "")
+                                  classpath.mkString(" -cp ", File.pathSeparator, "")
                                 else ""))

   /**
diff --git i/tools/src/test/scala/scala/scalanative/NIRCompiler.scala w/tools/src/test/scala/scala/scalanative/NIRCompiler.scala
index be4032f..49c1bcf 100644
--- i/tools/src/test/scala/scala/scalanative/NIRCompiler.scala
+++ w/tools/src/test/scala/scala/scalanative/NIRCompiler.scala
@@ -12,7 +12,7 @@ object NIRCompiler {
   private val classLoader = {
     val parts = sys
       .props("scalanative.testingcompiler.cp")
-      .split(":")
+      .split(File.pathSeparator)
       .map(new java.io.File(_))
       .filter(f => f.exists && f.getName.endsWith(".jar"))
       .map(_.toURI.toURL)
```

muxanick commented:
@jonas Perfect! It works now!

muxanick commented:
@jonas thanks Jonas, great changes, I will remove it from my PR as soon as it will be on master
